domain/pipeline/iterator_test.go

```diff
@@ -0,0 +1,280 @@
+		keys     []string
+		offset   int
+		expected []testdata
+	}{
+		{
+			name:     "Set offset to 0",
+			data:     []testdata{{key: "a", value: 1}, {key: "b", value: 2}, {key: "c", value: 3}},
+			keys:     []string{"a", "b", "c"},
+			offset:   0,
+			expected: []testdata{{key: "a", value: 1}, {key: "b", value: 2}, {key: "c", value: 3}},
+		},
+		{
+			name:     "Set offset to middle",
+			data:     []testdata{{key: "a", value: 1}, {key: "b", value: 2}, {key: "c", value: 3}},
+			keys:     []string{"a", "b", "c"},
+			offset:   1,
+			expected: []testdata{{key: "b", value: 2}, {key: "c", value: 3}},
+		},
+		{
+			name:     "Set offset to last element",
+			data:     []testdata{{key: "a", value: 1}, {key: "b", value: 2}, {key: "c", value: 3}},
+			keys:     []string{"a", "b", "c"},
+			offset:   2,
+			expected: []testdata{{key: "c", value: 3}},
+		},
+		{
+			name:   "Set offset beyond last element",
+			data:   []testdata{{key: "a", value: 1}, {key: "b", value: 2}, {key: "c", value: 3}},
+			keys:   []string{"a", "b", "c"},
+			offset: 3,
+		},
+		{
+			name:   "Set offset for empty map",
+			offset: 0,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			m := sync.Map{}
+			for _, v := range tt.data {
+				m.Store(v.key, v)
+			}
+
+			it := NewSyncMapIterator[string, testdata](&m, tt.keys)
+			it.SetOffset(tt.offset)
+
+			var result []testdata
+			for {
+				val, err := it.Next()
+				if err != nil {
+					break
+				}
+				result = append(result, val)
+			}
+
+			if !reflect.DeepEqual(result, tt.expected) {
+				t.Errorf("Iteration result after SetOffset(%d) = %v, want %v", tt.offset, result, tt.expected)
+			}
+
+			// Test that after full iteration, Next() returns false
+			_, err := it.Next()
+			if err == nil {
+				t.Errorf("Expected Next() to return err after full iteration")
+			}
+		})
+	}
+}
+
+func TestSyncMapIterator_HasNext(t *testing.T) {
+	tests := []struct {
+		name  string
+		keys  []string
+		index int
+		want  bool
+	}{
+		{
+			name:  "Empty iterator",
+			keys:  []string{},
+			index: 0,
+			want:  false,
+		},
+		{
+			name:  "Iterator with elements, at start",
+			keys:  []string{"a", "b", "c"},
+			index: 0,
+			want:  true,
+		},
+		{
+			name:  "Iterator with elements, in middle",
+			keys:  []string{"a", "b", "c"},
+			index: 1,
+			want:  true,
+		},
+		{
+			name:  "Iterator with elements, at last element",
+			keys:  []string{"a", "b", "c"},
+			index: 2,
+			want:  true,
+		},
+		{
+			name:  "Iterator with elements, past last element",
+			keys:  []string{"a", "b", "c"},
+			index: 3,
+			want:  false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			iterator := &SyncMapIterator[string, int]{
+				keys:  tt.keys,
+				index: tt.index,
+			}
+			if got := iterator.HasNext(); got != tt.want {
+				t.Errorf("SyncMapIterator.HasNext() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestSyncMapIterator_Reset(t *testing.T) {
```

Comment:
Let's also add a basic concurrency test where we utilize a sequence of iterator operations, arriving to the desired final state

Reply:
Calling iterator methods concurrently is not safe based on current implementation and [current usecase](https://github.com/osmosis-labs/sqs/pull/553/files#diff-d3c014d1dd38b58e90b490dd2aab2bf74aa545e71fa94ac7cdf6d317c4943434R513) does not include concurrent acesses of its methods. We could implement such improvements when needed in the future, but requiring to grab a lock for each method call atm would add extra overhead and unecesarry comlexity. Let me know if you would like to add this feature right now and I will follow-up with separate PR!